### DISCUSSION TITLE
다음 JavaScript 코드를 살펴보시고, 실행 결과를 예측해 주실 수 있나요?

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#71] [실행컨텍스트]
- 다음 JavaScript 코드를 살펴보시고, 실행 결과를 예측해 주실 수 있나요?
  ```js
  for (let i = 0; i < 3; i++) {
    setTimeout(function() {
      console.log(i);
    }, 1000 * i);
  }
  ```

  - 만약 let 대신 var를 사용한다면 결과가 어떻게 달라질까요? 그 차이점에 대해 설명해 주세요.
  - 이 코드에서 각 반복이 독립적인 i 값을 유지하는 것과 관련하여, JavaScript의 렉시컬 스코프에 대해 어떻게 이해하고 있나요?

**Comment by publdaze** [comment]
> - 다음 JavaScript 코드를 살펴보시고, 실행 결과를 예측해 주실 수 있나요?
>   ```js
>   for (let i = 0; i < 3; i++) {
>     setTimeout(function() {
>       console.log(i);
>     }, 1000 * i);
>   }
>   ```
- i는 let으로 선언되어 있음
- let 키워드로 선언된 변수는 블록 레벨 스코프를 따름
- 반복문에서 코드 블록을 반복 실행할 때마다 독자적인 렉시컬 환경을 생성하게 되는 것
- 따라서 setTimeout의 콜백에서 참조하고 있는 i가 상위 스코프인 각각의 렉시컬 환경에서 검색이 가능하게 되면서, 0,1,2초 후에 0,1,2가 순차적으로 출력됨
> - 만약 let 대신 var를 사용한다면 결과가 어떻게 달라질까요? 그 차이점에 대해 설명해 주세요.
- var는 블록 레벨 스코프가 아닌 함수 레벨 스코프를 따름
- 따라서 코드 블록이 반복 실행이 될 때 독자적인 렉시컬 환경을 생성하지 않고, i 식별자는 현재 코드에서 전역 렉시컬 환경에 저장됨
- 따라서 반복문이 종료될 당시의 i에 할당된 값은 `3`이고, 이후 실행되는 setTimeout 콜백에서는 전역 렉시컬 환경의 i, 즉 3이 할당되어있는 값을 참조하기 때문에 3,3,3이 출력됨
> - 이 코드에서 각 반복이 독립적인 i 값을 유지하는 것과 관련하여, JavaScript의 렉시컬 스코프에 대해 어떻게 이해하고 있나요?
- JavaScript는 렉시컬 스코프를 따름
- 렉시컬 스코프는 함수가 호출되는 위치가 아닌 정의되는 위치를 기반으로 스코프가 결정되는 것을 말함
- 위에서 setTimeout의 경우 콜백큐에서 대기하다가 실행컨텍스트 스택이 비워지는 걸 이벤트 루프에 의해 발견이 되면 실행컨텍스트 스택에 들어가며 실행이 됨
- 이때 호출되는 위치 기반이라면, 반복문이 끝나고, 실행 컨텍스트에서 비워졌기 때문에 반복문 하위에 있는 i 식별자를 참조하지 못하지만, 정의되는 위치 기반이기 때문에 상위 스코프의 i를 참조할 수 있어서 독립적인 i를 유지할 수 있는 것

**Comment by pipisebastian** [comment]
> 다음 JavaScript 코드를 살펴보시고, 실행 결과를 예측해 주실 수 있나요?
```js
for (let i = 0; i < 3; i++) {
  setTimeout(function() {
    console.log(i);
  }, 1000 * i);
}
```
> 이 코드에서 각 반복이 독립적인 i 값을 유지하는 것과 관련하여, JavaScript의 렉시컬 스코프에 대해 어떻게 이해하고 있나요?

- 0,1,2초 후, 0,1,2가 출력됩니다.
- let, const는 독자적인 렉시컬 환경을 생성합니다. 그렇기에 for문 반복문에서는, 순회를 할때마다 각자의 렉시컬 환경이 만들어집니다. 그렇기 때문에, setTimeOut의 콜백함수는, 스코프 체인을 통해 현재 console.log에서 참조하고 있는 i를 찾습니다. 이는 **함수가 선언된 위치**에서 상위 스코프인 for 반복문의 스코프를 확인(**렉시컬 스코프**)하고, 이때 반복문마다 만들어진 렉시컬 환경의 i를 찾게 됩니다.

> 만약 let 대신 var를 사용한다면 결과가 어떻게 달라질까요? 그 차이점에 대해 설명해 주세요.

- 0,1,2초 후, 3,3,3이 출력됩니다.
- var의 경우, 전역 렉시컬 환경에 저장됩니다. 그렇기에 반복문을 모두 순회했을 경우, i에는 3이 할당되며, setTimeout의 콜백함수에서는, 전역변수 i를 참조하므로, 3,3,3이 출력됩니다.
